gravitee-kubernetes-client/src/main/java/io/gravitee/gateway/services/kube/client/AbstractKubernetesResourceWatcher.java

```diff
@@ -0,0 +1,176 @@
+/**
+ * Copyright (C) 2015 The Gravitee team (http://gravitee.io)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.gravitee.gateway.services.kube.client;
+
+import io.gravitee.common.http.HttpHeaders;
+import io.gravitee.common.http.MediaType;
+import io.gravitee.gateway.services.kube.client.config.KubernetesConfig;
+import io.gravitee.gateway.services.kube.client.model.v1.Event;
+import io.reactivex.Observable;
+import io.reactivex.Single;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.core.http.HttpClientOptions;
+import io.vertx.core.http.WebSocketConnectOptions;
+import io.vertx.core.net.PemTrustOptions;
+import io.vertx.reactivex.core.Vertx;
+import io.vertx.reactivex.core.http.HttpClient;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.util.Assert;
+
+/**
+ * @author Kamiel Ahmadpour (kamiel.ahmadpour at graviteesource.com)
+ * @author GraviteeSource Team
+ * @since 3.9.11
+ */
+public abstract class AbstractKubernetesResourceWatcher implements KubernetesResourceWatcher {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(AbstractKubernetesResourceWatcher.class);
+
+    protected final Vertx vertx;
+    protected final KubernetesClient kubernetesClient;
+    protected final KubernetesConfig config;
+    protected final HttpClient httpClient;
+    protected Map<String, Watch> watchMap = new HashMap<>();
```

Comment:
final ?

Also, I'm wondering when this map is effectively clean. No `remove` on this map from this class - > potential memory leak?